model/host/hostutil.go

```diff
@@ -78,9 +78,19 @@ func (h *Host) CurlCommandWithRetry(settings *evergreen.Settings, numRetries, ma
 }
 
 func (h *Host) curlCommands(settings *evergreen.Settings, curlArgs string) []string {
+	var curlCmd string
+	if !settings.ServiceFlags.S3BinaryDownloadsDisabled && settings.HostInit.S3BaseURL != "" {
+		// Attempt to download the agent from S3, but fall back to downloading from
+		// the app server if it fails.
+		// Include -f to return an error code from curl if the HTTP request
+		// fails (e.g. it receives 403 Forbidden or 404 Not Found).
+		curlCmd = fmt.Sprintf("(curl -fLO '%s'%s || curl -LO '%s'%s)", h.S3ClientURL(settings), curlArgs, h.ClientURL(settings), curlArgs)
```

Comment:
the man page warns

> This method is not fail-safe and there are occasions where non-successful response codes will slip through, especially when authentication is involved (response codes 401 and 407).

is this a concern?

Reply:
I don't think so. We don't use authentication headers when retrieving objects from S3 because the URL has to have publicly read permissions. The error codes I would be concerned about are 403 Forbidden (i.e. you don't have permissions to access this object in S3) and 404 Not Found, which are covered by `-f`.